awx/main/access.py

```diff
@@ -2123,7 +2127,11 @@ class WorkflowJobAccess(BaseAccess):
     )
 
     def filtered_queryset(self):
-        return WorkflowJob.objects.filter(unified_job_template__in=UnifiedJobTemplate.accessible_pk_qs(self.user, 'read_role'))
+        return WorkflowJob.objects.filter(
+            Q(unified_job_template__in=UnifiedJobTemplate.accessible_pk_qs(self.user, 'read_role'))
+            | Q(created_by_id=self.user.id, is_bulk_job=True)
+            | Q(organization__in=Organization.objects.filter(Q(admin_role__members=self.user)), is_bulk_job=True)
```

Comment:
Yikes, this is going to be a tough pill to swallow. For the most part, _this_ query doesn't matter. When you go to the JOBS tab, the UI doesn't list workflow jobs. The UI lists `/api/v2/unified_jobs/` with a bunch of terribly scary filters (for scale) on it.

So right now, these jobs are probably not showing up in the UI for non-superusers. Ping @CFSNM 

How we move forward is not at all clear. The requirement that we would allow organization admins / auditors to see jobs launched by this mechanism (in their organization) is something we should have ideally pushed back on in early feature planning.

the UnifiedJob queryset is this:

https://github.com/ansible/awx/blob/17246ff1130de065557ece97b27b8defac952895/awx/main/access.py#L2501-L2509

It doesn't even contain the `created_by` filter, and importantly _I would push back on adding it_ for performance and stability reasons.

Reply:
Wait, maybe the `     | Q(organization__in=org_auditor_qs)` from the UnifiedJob query covers a lot of what you need here. And some of the update to the WorkflowJob queryset may just be bringing that to parity with the unified job queryset. However, this still leaves unsettled the question of created_by.

Reply:
Let me offer another way out - require `admin_role` for the `organization` field given in the bulk_launch endpoint. This would avoid the need to filter by `created_by`, but it has the downside of limiting who can use this feature.

If that's not workable, then we should question is workflow jobs are a suitable solution here. Users can still see the _jobs_ in any case, but if we let them do this without giving visibility to the workflow job, I fear this would be objectionable to the point that it's not sustainable.

Reply:
> So right now, these jobs are probably not showing up in the UI for non-superusers. Ping @CFSNM

@AlanCoding The UI maybe using a different query, because as a normal user in the UI I dont see the bulk job, but I do see them in API. 
```
[REST API](https://localhost:8043/api/) [Version 2](https://localhost:8043/api/v2/) [Workflow Job List](https://localhost:8043/api/v2/workflow_jobs/) 

Workflow Job List

GET /api/v2/workflow_jobs/?is_bulk_job=True

HTTP 200 OK
Allow: GET, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept
X-API-Node: awx_1
X-API-Product-Name: AWX
X-API-Product-Version: 21.10.3.dev60+g28f275ee25.d20230216
X-API-Query-Count: 19
X-API-Query-Time: 0.007s
X-API-Time: 0.035s

{
    "count": 10,
    "next": null,
    "previous": null,
    "results": [
```

Reply:
I see. I am not sure how do I move forward ? Currently you can't launch the bulk api via UI, is it fine if we say this is a API only feature and the job will be visible via API ?

At the same time I did some testing and wanted to post some numbers here:

- Created 10,000 (10k) users, assigned them admin access on one of the job templates
```
GET /api/v2/users/

HTTP 200 OK
Allow: GET, POST, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept
X-API-Node: awx_1
X-API-Product-Name: AWX
X-API-Product-Version: 21.10.3.dev60+g28f275ee25.d20230216
X-API-Query-Count: 124
X-API-Query-Time: 0.474s
X-API-Time: 1.376s

{
    "count": 10005,
    "next": "[/api/v2/users/?page=2](https://localhost:8043/api/v2/users/?page=2)",
    "previous": null,
    "results": [
```
- Ran multiple bulk jobs with different users (around 7k) jobs:
```
GET /api/v2/workflow_jobs/?is_bulk_job=True

HTTP 200 OK
Allow: GET, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept
X-API-Node: awx_1
X-API-Product-Name: AWX
X-API-Product-Version: 21.10.3.dev60+g28f275ee25.d20230216
X-API-Query-Count: 32
X-API-Query-Time: 0.039s
X-API-Time: 0.126s

{
    "count": 6839,
    "next": "[/api/v2/workflow_jobs/?is_bulk_job=True&page=2](https://localhost:8043/api/v2/workflow_jobs/?is_bulk_job=True&page=2)",
    "previous": null,
```
- logged in as one of those 10,000 (10k) user and tried fetching the workflow_jobs api endpoint:
```


GET /api/v2/workflow_jobs/

HTTP 200 OK
Allow: GET, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept
X-API-Node: awx_1
X-API-Product-Name: AWX
X-API-Product-Version: 21.10.3.dev60+g28f275ee25.d20230216
X-API-Query-Count: 14
X-API-Query-Time: 0.075s
X-API-Time: 0.145s

{
    "count": 11,
    "next": null,
    "previous": null,
```

API seems to very snappy, X-API-Query-Time and X-API-Time seems to be fine. I wonder if its because we have an index on `create_by_id` in the `main_unifiedjob` table ?

Reply:
It's very difficult to create the circumstances that people will see and complain about in the real world here. The most concerning endpoint is not this, but the unified jobs, because this gets _slammed_ by the UI, and the dangerous situation is where someone actually makes heavy use of the RBAC system so there is a lot of indirect role memberships, _in addition to_ the type-specific filters.

Right now, looking at the diff, `WorkflowJobAccess.filtered_queryset` is inconsistent with `UnifiedJob.filtered_queryset`. Where this has

```python
Q(organization__in=Organization.objects.filter(Q(admin_role__members=self.user)), is_bulk_job=True)
```

the `UnifiedJobAccess` queryset has

```python
Q(organization__in=org_auditor_qs)
```

Where `org_auditor_qs = Organization.objects.filter(Q(admin_role__members=self.user) | Q(auditor_role__members=self.user))`.

These really need to be the same filters, because a workflow job _is_ a unified job. The same items should appear in `/api/v2/workflow_jobs/` as what appears in `/api/v2/unified_jobs/`, otherwise it's just inconsistent and wrong.

So I agree that you should add this org admin / auditor filter, but you should add the same on as in the `UnifiedJobAccess` filter. And then my suggestion for this feature is to abandon the `created_by` filter.

Reply:
@AlanCoding I see what you are saying about `/api/v2/unified_jobs/` returning a different list than `/api/v2/workflow_jobs/`. However, in this world, there are no "broken" links. If you view in the UI the child jobs, you can then follow the link to the parent Bulk Job. and in the API you can follow the link for the workflow you spawned.

If consistency is paramount, we could make a second query in the UnifiedJobsAccess like we do for some of the other conditions
```
diff --git a/awx/main/access.py b/awx/main/access.py
index f5eda98ad4..a999b4be63 100644
--- a/awx/main/access.py
+++ b/awx/main/access.py
@@ -2500,11 +2500,13 @@ class UnifiedJobAccess(BaseAccess):
     def filtered_queryset(self):
         inv_pk_qs = Inventory._accessible_pk_qs(Inventory, self.user, 'read_role')
         org_auditor_qs = Organization.objects.filter(Q(admin_role__members=self.user) | Q(auditor_role__members=self.user))
+        self_bulk_job_ids = WorkflowJob.objects.filter(Q(created_by_id=self.user.id, is_bulk_job=True)).values_list('id', flat=True)
         qs = self.model.objects.filter(
             Q(unified_job_template_id__in=UnifiedJobTemplate.accessible_pk_qs(self.user, 'read_role'))
             | Q(inventoryupdate__inventory_source__inventory__id__in=inv_pk_qs)
             | Q(adhoccommand__inventory__id__in=inv_pk_qs)
             | Q(organization__in=org_auditor_qs)
+            | Q(id__in=self_bulk_job_ids)
         )
         return qs
```

Reply:
Let me show you what that will look like when it generates the SQL. Persona is an org admin here.

https://gist.github.com/AlanCoding/9cacdb6e5770811194992ccf88b573bd

What does your proposed change add? The following, in sort of the main list of OR clauses.

```sql
        OR "main_unifiedjob"."id" IN
          (SELECT U0."unifiedjob_ptr_id"
           FROM "main_workflowjob" U0
           INNER JOIN "main_unifiedjob" U1 ON (U0."unifiedjob_ptr_id" = U1."id")
           WHERE (U1."created_by_id" = 1097
                  AND U0."is_bulk_job")))
```

Big picture - avoiding queries like this is _why_ the RBAC system was created in the first place. These quickly become unmaintainable, and people will lose the ability to use the app due to complex permissions and a large number of objects.

This isn't a linear cost, either. Every new condition is tremendously risky.

We shouldn't be doing these conditions at all. There should be _one_ filter implemented by the backend to filter with roles, plus what the UI filters by. We should not have added the workflow approval condition, and that doesn't justify adding a new condition now.

Reply:
ok talked to @AlanCoding and we came up with https://github.com/jainnikhil30/awx/pull/15 as a resolution to these problems